server/command_list.go

```diff
@@ -28,23 +27,43 @@ func (p *Plugin) runListCommand(args []string, extra *model.CommandArgs) (*model
 }
 
 func (p *Plugin) getUpdatedInstallsForUser(userID string) ([]*Installation, error) {
-	installsForUser, err := p.getInstallationsForUser(userID)
+	pluginInstalls, err := p.getInstallationsForUser(userID)
 	if err != nil {
 		return nil, err
 	}
 
-	var updatedInstall *cloud.Installation
-	for _, install := range installsForUser {
-		updatedInstall, err = p.cloudClient.GetInstallation(install.ID)
-		if err != nil {
-			return nil, errors.Wrapf(err, "could not get updated installation %s", install.ID)
-		}
-		if updatedInstall == nil {
-			return nil, fmt.Errorf("could not find installation %s", install.ID)
+	cloudInstalls, err := p.cloudClient.GetInstallations(&cloud.GetInstallationsRequest{
+		OwnerID:        userID,
+		IncludeDeleted: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, cloudInstall := range cloudInstalls {
+		for j, pluginInstall := range pluginInstalls {
```

Comment:
Hmm, I could see this nested loop leading to long-term performance hits. Here is what I am am thinking.

The `cloudInstall` slice will slowly get larger and larger as the user creates more cloud installations. I know that it is unlikely that this will become a real issue, but considering this slice will only grow (in that even deleted installations are always returned) I think we can try something else.

We could keep logic similar to the original code where it assumes the installation is not deleted and asks for it. If it is returned as expected, we proceed as normal. If it isn't returned correctly, then make a follow-up query to make sure it was deleted. If it was deleted according to the provisioner, we cleanup the plugin store.

What do you think of this?

Reply:
> We could keep logic similar to the original code where it assumes the installation is not deleted and asks for it. If it is returned as expected, we proceed as normal. If it isn't returned correctly, then make a follow-up query to make sure it was deleted. If it was deleted according to the provisioner, we cleanup the plugin store.

I think we can do that for now but it still does solve the problem with a large data set.  Also, if that plan is to never remove records related to deleted instances, I think we should also consider to fetch this data in chunks instead of the whole thing.